services/venice-router/src/main/java/com/linkedin/venice/router/api/DictionaryRetrievalService.java

```diff
@@ -216,6 +216,16 @@ public DictionaryRetrievalService(
     this.dictionaryRetrieverThread = new Thread(runnable);
   }
 
+  public byte[] getDictionaryForKafkaTopic(String kafkaTopic) {
```

Comment:
I wonder if this function should check whether there is already a future for downloading this dictionary in `downloadingDictionaryFutures`, and if there is, perhaps wait on that instead of kicking off a new one?

I wonder If we may end up with a thundering herd problem otherwise...

On the other hand, if the one future that's in flight is unlucky and hits an unhealthy server that hangs a while before responding, then we could end up having many requests tied up behind the same future, which may be not much better than the thundering herd.

I'm not sure what is better. What do you think?

Reply:
Given we select random host on each request (`getOnlineInstance`) the chances for hitting a same server would be low. 
We should definitely check the existing future queue and if found wait that future. If that future times out we can kick off another future (which hopefully will select another host)?